### PR TITLE
Patch CTID

### PR DESCRIPTION
**Pull request by dangell7**
## High Level Overview of Change

Fix a number of issues involved with CTID. 

1. CTID is not present on all RPC `tx` transactions.
2. rpcWRONG_NETWORK was missing in the `ErrorCodes.cpp`

- FIx https://github.com/XRPLF/rippled/issues/4735

### Context of Change

- Initial CTID support added in: #4418
- [XLS-37 specification](https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0037d-concise-transaction-identifier-ctid#readme)

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [x] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

- [x] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

### CODE REVIEW

**Comment by cindyyan317** [comment]
I ran this PR locally. I can see there is ctid field in subscription and account_tx now. However, i can not see ctid in tx response any more.
My input:
'''
{
    "method": "tx",
    "params":[{ "transaction": "CCBBD2F5CEE90FA7A2ED13A7A343D015F57950B562185734E125B7BE740F2428"}]
}
'''
output:
Screenshot 2023-10-03 at 10 17 41

"ctid" is available for the same transaction from account_tx response.

**Comment by dangell7** [comment]
> I ran this PR locally. I can see there is ctid field in subscription and account_tx now. However, i can not see ctid in tx response any more. My input: ''' { "method": "tx", "params":[{ "transaction": "CCBBD2F5CEE90FA7A2ED13A7A343D015F57950B562185734E125B7BE740F2428"}] } ''' output: Screenshot 2023-10-03 at 10 17 41
> 
> "ctid" is available for the same transaction from account_tx response.

Whats the networkID? Can you show me the json for the response that was correct?

**Comment by dangell7** [comment]
> > > I ran this PR locally. I can see there is ctid field in subscription and account_tx now. However, i can not see ctid in tx response any more. My input: ''' { "method": "tx", "params":[{ "transaction": "CCBBD2F5CEE90FA7A2ED13A7A343D015F57950B562185734E125B7BE740F2428"}] } ''' output: Screenshot 2023-10-03 at 10 17 41
> > > "ctid" is available for the same transaction from account_tx response.
> > 
> > 
> > Whats the networkID? Can you show me the json for the response that was correct?
> 
> I should mention I am on devnet. I think the correct one should contain the "ctid", right?

It should... I will review this again to make sure.

**Comment by dangell7** [comment]
@intelliot I believe devent was reset to a networkID < 1024 therefore it did not have a ctid. @cindyyan317 can you please recheck the devnet again with your locally running PR?

**Comment by cindyyan317** [comment]
> @cindyyan317 can you please recheck the devnet again with your locally running PR?

Hi @dangell7 . I can see the ctid field from tx now. I guess ctid is supposed to be everywhere? I don't see it in "ledger" and "transaction_entry"

>@intelliot I believe devent was reset to a networkID < 1024 therefore it did not have a ctid. 
What does it mean?  It should not have ctid when networdID < 1024? devnet's networkId has been always 2 , right?

**Comment by intelliot** [comment]
Yes, I think Devnet's NetworkID has always been `2`.

@dangell7 , I think it should be possible (and would be ideal) to always return `ctid` with validated transactions, for any NetworkID (even if it's <= 1024).

**Comment by dangell7** [comment]
> Yes, I think Devnet's NetworkID has always been `2`.
> 
> @dangell7 , I think it should be possible (and would be ideal) to always return `ctid` with validated transactions, for any NetworkID (even if it's <= 1024).

You're right and I will look at the issue deeper this weekend. I don't like peering from my work computer. I have an env now to spin up a peer easier to test. 

I will also check on the CTID on txs <= 1024. I think I misspoke.

**Comment by dangell7** [comment]
@intelliot moving this to draft while I write more tests. Can confirm ctid is returned when network id <= 1024.

**Comment by dangell7** [comment]
@intelliot I'm opening this back up for review. I see no issues

**Review by scottschurr — COMMENTED** [review]
I left a number of comments for your consideration.  I'll highlight the important ones.

1. I think the signature of `RPC::encodeCTID()` is wrong.  It should accept all three parameters as `uint32_t`.  That will allow all of the parameter validation to be localized in `RPC::encodeCTID()` and not spread around the code base.  That means that values passed to `RPC::encodeCTID()` should not be prematurely truncated to `uint16_t`.
2. As mentioned above, once the signature of `RPC::encodeCTID()` is fixed, all of the parameter validation should take place there.
3. There is currently no place in the code base that describes what the acronym CTID means.  It seems like that explanation ought to be _someplace_.  I had to cast about for a while to try and figure out what CTID meant, so I could figure out what was going on.
4. It feels like the limit on how large an acceptable network ID can be is cutting off a little too early.  Shouldn't it be able to go all the way to 0xFFFF?
5. A couple of the unit tests look like they may be almost duplicates, other than the network ID value.  I think things will be easier to review and maintain if those almost-duplicated tests are collapsed into a single `for` loop.  If I'm wrong about being able to collapse those tests, there should be comments highlighting the differences.  I could not find the differences through casual inspection.

I'm open to being wrong on any of these comments, since I was not tracking initial development of Compact Transaction IDs.  Please feel free to push back on them.

On the off chance that my comments are not far off base, I have a commit with all of the changes I'm suggesting.  You'll find it as the top-most commit here: https://github.com/scottschurr/rippled/commits/dangell7-patch-ctid  Feel free to use that as an example.

Please be cautious about cherry-picking that commit, however.  If you choose to cherry-pick it you should review it carefully in case I took too many liberties, particularly in the testing.

Thanks for your submission.  In particular, I really appreciate that you included unit tests.

**Comment by intelliot** [comment]
@dangell7 - I just updated your PR description to use the template. Please feel free to edit / add more. Also, consider the suggestions from scottschurr above, and comment here when you are done.

**Comment by dangell7** [comment]
@scottschurr Thank you very much for the review. I looked over the code and everything looks fine. I'm going to run the tests and build a version to test locally. 

Thank you

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 1]
```diff
@@ -3110,6 +3111,21 @@ NetworkOPsImp::transJson(
             jvObj[jss::meta], *ledger, transaction, meta->get());
     }
 
+    // add CTID where the needed data for it exists
+    if (auto const& lookup = ledger->txRead(transaction.getTransactionID());
+        lookup.second && lookup.second->isFieldPresent(sfTransactionIndex))
+    {
+        uint32_t txnSeq = lookup.second->getFieldU32(sfTransactionIndex);
```
Comment (scottschurr) [inline src/ripple/app/misc/NetworkOPs.cpp:3118]: `txnSeq` could be `const`

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 2]
```diff
@@ -3110,6 +3111,21 @@ NetworkOPsImp::transJson(
             jvObj[jss::meta], *ledger, transaction, meta->get());
     }
 
+    // add CTID where the needed data for it exists
+    if (auto const& lookup = ledger->txRead(transaction.getTransactionID());
```
Comment (scottschurr) [inline src/ripple/app/misc/NetworkOPs.cpp:3115]: `transaction->getTransactionID()` should be `transaction->getTransactionID()`

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 3]
```diff
@@ -3110,6 +3111,21 @@ NetworkOPsImp::transJson(
...
+        lookup.second && lookup.second->isFieldPresent(sfTransactionIndex))
+    {
+        uint32_t txnSeq = lookup.second->getFieldU32(sfTransactionIndex);
+        uint32_t netID = app_.config().NETWORK_ID;
+        if (transaction.isFieldPresent(sfNetworkID))
```
Comment (scottschurr) [inline src/ripple/app/misc/NetworkOPs.cpp:3120]: `transaction.isFieldPresent` should be `transaction->isFieldPresent`

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 4]
```diff
@@ -3110,6 +3111,21 @@ NetworkOPsImp::transJson(
...
+            netID = transaction.getFieldU32(sfNetworkID);
```
Comment (scottschurr) [inline src/ripple/app/misc/NetworkOPs.cpp:3121]: `transaction.getFieldU32` should be `transaction->getFieldU32`

File: src/ripple/app/misc/impl/Transaction.cpp [hunk 5]
```diff
@@ -180,6 +189,20 @@ Transaction::getJson(JsonOptions options, bool binary) const
...
+
+        // compute outgoing CTID
+        // override local network id if it's explicitly in the txn
+        std::optional netID = mNetworkID;
+        if (mTransaction->isFieldPresent(sfNetworkID))
+            netID = mTransaction->getFieldU32(sfNetworkID);
+
+        if (mTxnSeq && netID)
+        {
+            std::optional<std::string> ctid =
```
Comment (scottschurr) [inline src/ripple/app/misc/impl/Transaction.cpp:201]: `ctid` could be `const`.

**Review by scottschurr — CHANGES_REQUESTED** [review]
Well, it turns out that the changes I suggested included some bugs.  I'd like to get those fixed, so I'm asking for one more round of changes.  Here's the summary:

1. There are places where I wrote "Compact" in comments where I should have written "Concise".  My fault.  Could you fix those please?
2. It looks like the most recent merge commit was not quite up to snuff.  In Transaction.cpp there are a few places where `transaction.` need to become `transaction ->`.
3. During the review I spotted a few more locals that could be `const`.

I think issue 2 has to be fixed, or the code won't build.  Issue 1 really _should_ be fixed.  So you might as well patch up the `const`s while you're in there.

Again, sorry about introducing bugs.

**Comment by dangell7** [comment]
@scottschurr Thanks again for your update. I have fixed all of the issues. However, we are getting some weird behavior from this. Its sometimes on the tx and sometimes not. We are chasing this down now. 

I believe more tests need to be written for this or a bug needs to be fixed. As soon as we can replicate the issue I will fix this.

**Comment by intelliot** [comment]
Given 

> However, we are getting some weird behavior from this. Its sometimes on the tx and sometimes not. We are chasing this down now.

Could you `Convert to draft` this PR so we know it isn't ready for merging right now? 

Of course, when you do think it may be ready, please mark it as "ready for review".

**Comment by intelliot** [comment]
Internal tracker: RPFC-82

**Comment by intelliot** [comment]
note: @dangell7 indicated he will push changes tomorrow.

**Comment by intelliot** [comment]
- awaiting update or comment from @dangell7

**Comment by intelliot** [comment]
- there is some missing unit test coverage (see above)
- @dangell7 to let us know when this is, from his perspective, ready to merge

**Review by scottschurr — APPROVED** [review]
:+1: Looks good to me.

**Comment by dangell7** [comment]
@intelliot working on the codecov

File: src/ripple/app/misc/impl/AccountTxPaging.cpp [hunk 6]
```diff
@@ -44,12 +44,19 @@ convertBlobsToTxResult(
...
     auto tr = std::make_shared<Transaction>(txn, reason, app);
 
-    tr->setStatus(Transaction::sqlTransactionStatus(status));
-    tr->setLedger(ledger_index);
-
     auto metaset =
         std::make_shared<TxMeta>(tr->getID(), tr->getLedger(), rawMeta);
 
+    // if properly formed meta is available we can use it to generate ctid
+    if (metaset->getAsObject().isFieldPresent(sfTransactionIndex))
```
Comment (dangell7) [inline src/ripple/app/misc/impl/AccountTxPaging.cpp:51]: @scottschurr Can you look at this? Wouldn't sfTransactionIndex always exist? If I do `account_tx` on a non validated ledger I get "Ledger not validated." `SetStatus` values are optional. So can we just call this every time?

Comment (dangell7) [reply src/ripple/app/misc/impl/AccountTxPaging.cpp:51]: Or how do I test `account_tx` with a tx that doesnt have a sfTransactionIndex?

Comment (scottschurr) [reply src/ripple/app/misc/impl/AccountTxPaging.cpp:51]: @dangell7, thanks for asking.  I can't immediately think of a situation where you could retrieve metadata for a transaction from a ledger and not have `sfTransactionIndex` filled in.  However I prefer the conservative approach of verifying that the field is present before accessing it.  Bugs happen.  Doing checks like this can prevent a bug elsewhere from turning catastrophic.

Taking a step back, we ran for a few years without code coverage as part of CI. And, thanks to some great work by good folks, we started getting code coverage reports again recently.  But we're still learning how to best apply the tool.  There will be code that can't be exercised by reasonable tests (testing `asserts` anyone?).  We have a new microscope and we've discovered that there are tiny mites living on our skin.  Over time we'll learn to relax and know we don't have to check that each individual mite is harmless.

Having said that, if I were bound and determined to exercise the `else` condition here, I'd doing it by calling the `convertBlobsToTxResult()` function directly.  That would allow me to pass in whatever I wanted in the `rawMeta` argument.  This test would pass in `rawMeta` that omitted an `sfTransactionIndex`.  I'd guess that getting that test in place might take a day or so of effort.

Comment (dangell7) [reply src/ripple/app/misc/impl/AccountTxPaging.cpp:51]: Awesome. Thank you for this response. I do like the code coverage stuff. I will take a look at how to get some more coverage on this. `convertBlobsToTxResult` sounds pretty simple. Thanks

**Comment by intelliot** [comment]
@dangell7 how would you like this to move forward?

**Review by seelabs — APPROVED** [review]
:+1: The code looks good to me. @dangell7 is this ready to merge or are there still things that need to be resolved?

**Comment by dangell7** [comment]
@seelabs Yeah this is ready. I tried to test the `convertBlobsToTxResult` but was unable to build a good testcase without rewriting the function.

**Comment by intelliot** [comment]
`seelabs` confirmed that this still needs perf testing. cc @sophiax851

**Comment by sophiax851** [comment]
I tried to build a rippled with this PR to run on the devnet's full history node, but since this PR was based off the 2.2.0-rc1, so the build is getting amendment blocked. @dangell7 can you please rebase off the master branch so I can build it for 2.2.0? Thanks.

**Comment by sgramkumar** [comment]
@dangell7  running a test to use ctid in lowercase seems to be breaking the test

```
2025-02-05 10:33:51 (                    rippled.py:1769) INFO     {'method': 'tx', 'params': [{'ctid': 'c000012300000007', 'Fee': 1000}]}
2025-02-05 10:33:51 (                    rippled.py:1787) INFO       Response status: error
2025-02-05 10:33:51 (                  ctid_test.py:130 ) INFO     {'result': {'error': 'invalidParams', 'error_code': 31, 'error_message': 'Invalid parameters.', 'request': {'Fee': 1000, 'command': 'tx', 'ctid': 'c000012300000007'}, 'status': 'error'}}
```

**Same works for upper case ctid**
```
2025-02-05 11:05:43 (                    rippled.py:1769) INFO     {'method': 'tx', 'params': [{'ctid': 'C000001F00000007', 'Fee': 1000}]}
2025-02-05 11:05:43 (                    rippled.py:1787) INFO       Response status: success
2025-02-05 11:05:43 (                  ctid_test.py:130 ) INFO     {'result': {'Account': 'rPD1i8CacpThw8ysjWKhquX1R6BvfeXdjV', 'Amount': '1000', 'DeliverMax': '1000', 'Destination': 'rBEAWFjdLepAsmqSBNHdN7ph9sdPmKR8Ac', 'Fee': '1000', 'Flags': 2147483648, 'Sequence': 16, 'SigningPubKey': '02538BA5CC8BF439B8411C21B100283008AED149E27BE46FA0D683288FA5CA8B06', 'TransactionType': 'Payment', 'TxnSignature': '30450221009FAB43B32C16FEF2FB362F3063A7C1868B9636EE3329C2EB444E65071761A1F802205B44603505F0DCC00E03A2E426FAAB673FA320F2A2A19716AC0F3E63A948D4C1', 'ctid': 'C000001F00000007', 'date': 792097568, 'hash': '659A69CC7F299156EAAB382D5A5C77047F1D0D57AB3C9541911D39927DBF7ECD', 'inLedger': 31, 'ledger_index': 31, 'meta': {'AffectedNodes': [{'ModifiedNode': {'FinalFields': {'Account': 'rPD1i8CacpThw8ysjWKhquX1R6BvfeXdjV', 'Balance': '1999996000', 'Flags': 0, 'OwnerCount': 0, 'Sequence': 17}, 'LedgerEntryType': 'AccountRoot', 'LedgerIndex': '7240AF71D70203DBA8811044DD7668F253784B26E67E1ACCA21B5C55FAA4F900', 'PreviousFields': {'Balance': '1999998000', 'Sequence': 16}, 'PreviousTxnID': 'FFC2142D82425257E83AAD939039DD9E551D6523538F47FD7CE623AB294F9F94', 'PreviousTxnLgrSeq': 27}}, {'ModifiedNode': {'FinalFields': {'Account': 'rBEAWFjdLepAsmqSBNHdN7ph9sdPmKR8Ac', 'Balance': '2000002000', 'Flags': 0, 'OwnerCount': 0, 'Sequence': 22}, 'LedgerEntryType': 'AccountRoot', 'LedgerIndex': 'E2FAC88A9631B61F75DBA34FA558E0D384D740FA660A5C223F8999F33520687E', 'PreviousFields': {'Balance': '2000001000'}, 'PreviousTxnID': 'FFC2142D82425257E83AAD939039DD9E551D6523538F47FD7CE623AB294F9F94', 'PreviousTxnLgrSeq': 27}}], 'TransactionIndex': 0, 'TransactionResult': 'tesSUCCESS', 'delivered_amount': '1000'}, 'status': 'success', 'validated': True}}

```




Also, this PR needs to be rebased. Still hitting `server blocked`
> 2025-Feb-05 18:02:39.486379 UTC LedgerMaster:ERR One or more unsupported amendments activated: server blocked.

**Comment by dangell7** [comment]
Commit Message:

```
fix: CTID improvements

Update RPC handlers and transaction processing logic to correctly
generate and include Concise Transaction Identifiers (CTIDs) in
responses when appropriate. Add a new RPC error code rpcWRONG_NETWORK
to handle cases where a CTID is submitted to a node running on a
different network. Improve CTID validation logic and expand unit
tests to cover network ID boundary conditions and error scenarios.
```

**Comment by sophiax851** [comment]
I'm traveling today, I'll try to run a load and see if there's any regression comparing with a same load on 2.4.0.

**Comment by dangell7** [comment]
> Hi @dangell7, Is this branch based off the 2.4.0? I've done some testing with the following Load Scenario: **A TX load of 500 TPS combined with a total of ~150 TPS of mixed transaction load injected on the 4 client nodes of a private XRPL network consisting of 5 validators and 4 CH nodes.** A set of comparison tests were run on the same network running rippled 2.4.0 vs. rippled built from this PR. The results are showing about 22% of latency improvement with the TX RPC calls and a 13% of improvement with the Submit RPC calls. There's also a slight improvement with Ledger Consensus latency. I'm very curious about the source of the improvements if this branch was indeed built off rippled_2.4.0. I will run another test on 2.4.0 just to verify the results are consistent. cc: @ximinez

Well thats very interesting. I just updated the branch just in case but I'm pretty sure I rebased off 2.4.0 last week.

**Comment by sophiax851** [comment]
I think your previous branch was based off 2.4.0: 
"ubuntu@sophia-vl2:/opt/ripple/bin$ ./rippled_patchCTID --version
rippled version 2.4.0
Git commit hash: 08a3775f136a4178ec53c1127be42bf23e6f89d3
Git build branch: patch-ctid"
I reran the same time on 2.4.0 again, the difference was almost all gone, then I took a look of the graphs, looks like the previous test on 2.4.0 had some network issue with the environment at the end of the test, which skewed the results, see below:
rippled2 4 0_responseTime

I think we are all good, I'd sign off on this PR. Thanks for your work!!

**Comment by mvadari** [comment]
@bthomee this looks good to go

**Review by godexsoft — APPROVED** [review]
No objections from Clio side. Tx handler's logic looks fine.

**Comment by bthomee** [comment]
@dangell7 can you please update the branch? I'll merge it right afterwards.
